### PR TITLE
HybridStrategy (ParameterServerStrategy + CollectiveAllReduceStrategy)

### PR DESCRIPTION
**Pull request by elzino**
Implemented HybridStrategy by mixing ParameterServerStrategy and CollectiveAllReduceStrategy

**Major feature:**
  - Only in embedding layer case, HybridStrategy creates variable on PS. And in other cases, it creates Mirrored variables.
  - HybridStrategy chooses the different ways to reduce depending on the variable type.
  - ParameterServerStrategy trains model asynchronously. So i added codes to synchronize gradient updates in optimizer_v2.py file.

**Minor feature:**
- added trainable property and value property to AggregatingVariable

### CODE REVIEW

File: tensorflow/python/distribute/input_lib.py [hunk 1]
```diff
@@ -534,10 +534,15 @@ def __init__(self,
     if input_context:
       # Between-graph where we rely on the input_context for sharding
       assert input_workers.num_workers == 1
```
Comment (sj6077) [inline tensorflow/python/distribute/input_lib.py:536]: num_workers가 1이라는 건 worker 1개 = GPU 1개를 쓴다는 말인가요?

File: tensorflow/python/distribute/input_lib.py [hunk 2]
```diff
@@ -534,10 +534,15 @@ def __init__(self,
...
-      dataset = input_ops.auto_shard_dataset(dataset,
-                                             input_context.num_input_pipelines,
-                                             input_context.input_pipeline_id)
-      self._cloned_datasets.append(dataset)
+      worker = input_workers.worker_devices[0]
```
Comment (sj6077) [inline tensorflow/python/distribute/input_lib.py:537]: auto_shard_dataset을 쓰지 못하는 이유는 뭔가요?

File: tensorflow/python/keras/optimizer_v2/optimizer_v2.py [hunk 3]
```diff
@@ -457,6 +460,49 @@ def apply_grad_to_update_var(var, grad):
 
       apply_kwargs = {}
       if isinstance(grad, ops.IndexedSlices):
+        if isinstance(distribution, hybrid_strategy.HybridStrategy):
```
Comment (sj6077) [inline tensorflow/python/keras/optimizer_v2/optimizer_v2.py:463]: gradient accumulate하는 부분에서 이미 reduced_grads를 update하는 것 같은데, 그러면 각 worker가 이미 average한 grad를 다시 average하려고 시도 하는 건 아닌가요? reduced_grads가 sparse인 경우에는 아무 것도 안하고 그냥 원래 grad를 그대로 리턴하나요?

File: tensorflow/python/distribute/input_lib.py (hunk 1 above)
Comment (elzino) [reply tensorflow/python/distribute/input_lib.py:536]: 아닙니다 저기서 worker는 input_worker여서 data를 읽어오는 worker를 말합니다 GPU와는 관계 없습니다.
추가로 현재 tf2.1에서 strategy(분산학습)을 사용하면 항상 input_worker 갯수가 1개로 설정되어 있습니다

File: tensorflow/python/distribute/input_lib.py (hunk 2 above)
Comment (elzino) [reply tensorflow/python/distribute/input_lib.py:537]: 저 코드를 그대로 사용하면 아래 같은 오류가 납니다 
image

Comment (elzino) [reply tensorflow/python/distribute/input_lib.py]: distribute.replicate 함수를 사용하면 해결되서 이 코드를 추가했습니다.

Comment (elzino) [reply tensorflow/python/distribute/input_lib.py:537]: distribute.replicate 함수를 사용하면 해결되서 이 코드를 추가했습니다.

File: tensorflow/python/keras/optimizer_v2/optimizer_v2.py (hunk 3 above)
Comment (elzino) [reply tensorflow/python/keras/optimizer_v2/optimizer_v2.py:463]: argument를 넘겨서 average안하고 SUM을 합니다
사실 질문을 잘 이해 못해서 조금 더 자세하게 설명해주실 수 있으신가요?

Comment (sj6077) [reply tensorflow/python/keras/optimizer_v2/optimizer_v2.py:463]: accumulator가 하는 일이 w1의 grad1과 w2의 grad2를 더해서 평균 값을 내주는데, 여기 코드에서 보면 먼저 grad1+grad2를 더해서 w1과 w2가  reduced_grads로 가지고 있다가, accumulator에 보내는데요(그럼 accumulator에서 reduced_grads + reduced_grads / 2가 수행되겠죠?). 만약 SUM을 하는 부분이 실제로 sparse grad에 대해서도 수행되는 부분이라면 불필요하게 통신이 2번 일어나는 일이 발생합니다. 그래서 실제로 SUM을 수행하는지 안하는지에 대한 질문 이었습니다.

File: tensorflow/python/distribute/input_lib.py (hunk 2 above)
Comment (sj6077) [reply tensorflow/python/distribute/input_lib.py:537]: 기존에도 여러 worker 여러 GPU가 서로 data sharding을 할 수 있게 동작하던 코드가 아닌가요? 왜 갑자기 에러가 나는 건가요?

File: tensorflow/python/keras/optimizer_v2/optimizer_v2.py (hunk 3 above)
Comment (elzino) [reply tensorflow/python/keras/optimizer_v2/optimizer_v2.py:463]: reduced_grads는 한 worker에서 여러 GPU에서 계산된 gradient들이 더해지지 않고 concat되어 있는 상태입니다. local aggregation이라고 생각하시면 편하실 거 같습니다.
여러 worker간 구해진 gradient들을 합치기 위해서 Acuumulator에 보내고 합치는 일을 합니다!

File: tensorflow/python/distribute/input_lib.py (hunk 2 above)
Comment (elzino) [reply tensorflow/python/distribute/input_lib.py:537]: 기존 방식에서는 CollectiveAllReduce만 여러 worker간에 data sharding을 진행해서 PS부분을 추가하는 것에 대한 고려가 없이 짜여져 저희가 PS를 추가하려다 보니 오류가 난다고 생각합니다.

Comment (elzino) [reply tensorflow/python/distribute/input_lib.py:537]: 그리고 auto_shard_dataset은 코드 542번째 줄을 보시면 사용하고 있습니다!
